docs/source/rabbitmq-tutorial/7-publisher-confirms.rst

```diff
@@ -0,0 +1,146 @@
+waiting synchronously for its confirmation:
+
+.. literalinclude:: examples/7-publisher-confirms/publish_individually.py
+   :language: python
+   :lines: 22-30
+
+In the previous example we publish a message as usual and wait for its confirmation with the `asyncio.wait_for` method.
+The method returns as soon as the message has been confirmed.
+If the message is not confirmed within the timeout or if it is nack-ed (meaning the broker could not take care of it for
+some reason), the method will throw an exception.
+The handling of the exception usually consists in logging an error message and/or retrying to send the message.
+
+If a timeout isn't requied the `await` keyword can be used instead directly as before.
+
+Different client libraries have different ways to synchronously deal with publisher confirms, so make sure to read
+carefully the documentation of the client you are using.
+
+This technique is very straightforward but also has a major drawback: it significantly slows down publishing, as the
+confirmation of a message blocks the publishing of all subsequent messages.
+This approach is not going to deliver throughput of more than a few hundreds of published messages per second.
+Nevertheless, this can be good enough for some applications.
+
+..
+   TODO: Do we need to specify, that wait_for cancels the awaitables waiting for publisher confirmation when the timeout
+   occurs?
+
+Strategy #2: Publishing Messages in Batches
++++++++++++++++++++++++++++++++++++++++++++
+
+To improve upon our previous example, we can publish a batch of messages and wait for this whole batch to be confirmed.
+The following example uses a batch of 100:
+
+.. literalinclude:: examples/7-publisher-confirms/publish_batches.py
+   :language: python
+   :lines: 22-45
+
+Waiting for a batch of messages to be confirmed improves throughput drastically over waiting for a confirm for individual
+message (up to 20-30 times with a remote RabbitMQ node).
+One drawback is that we do not know exactly what went wrong in case of failure, so we may have to keep a whole batch in memory
+to log something meaningful or to re-publish the messages.
+And this solution is still synchronous, so it blocks the publishing of messages.
+
+Strategy #3: Handling Publisher Confirms Asynchronously
++++++++++++++++++++++++++++++++++++++++++++++++++++++++
+
+The broker confirms published messages asynchronously, one just needs to register a callback on the client to be notified of these confirms:
+
+.. literalinclude:: examples/7-publisher-confirms/publish_asynchronously.py
+   :language: python
+   :pyobject: handle_confirm
+
+.. literalinclude:: examples/7-publisher-confirms/publish_asynchronously.py
+   :language: python
+   :lines: 32-40
+
+The `.result()` method will either return a `aiormq.abc.ConfirmationFrameType` for confirmed messages
+or raise an Exception for nack-ed messages (messages that can be considered lost by the broker).
+
+..
+   TODO: Which exceptions can be raised?
```

Comment:
I tried to look where the Future is resolved and found [this code in `aiormq`](https://github.com/mosquito/aiormq/blob/f04f0cf7f1972e14abd4fefc90218770f335e239/aiormq/channel.py#L348-L370).

To me it looks like it only ever resolved to `Basic.Ack`  and raises a `DeliveryError` for everything else.
Personally I actually think this would be a good thing, because iterating the responses to check for Ack/Nack/Reject is annoying and repetitive. This way a simple `asyncio.gather` can be used and will throw if anything is not `Ack`ed. This behavior also seems to match the `wait_for_pending_acks`/`WaitForConfirmsOrDie` method from the tutorial and other libraries.

I didn't test what actually happens, though.

Reply:
I figured, that the only way to get more information about the message to which the Confirmation belongs, without changing `aio_pika`, is to use a `ContextVar`, so I added that as a hint to the documentation.